xds/internal/balancer/clusterresolver/resource_resolver_eds.go

```diff
@@ -64,31 +61,44 @@ func (er *edsDiscoveryMechanism) stop() {
 	er.cancelWatch()
 }
 
-func (er *edsDiscoveryMechanism) handleEndpointsUpdate(update xdsresource.EndpointsUpdate, err error) {
-	if er.stopped.HasFired() {
-		return
+// newEDSResolver returns an implementation of the endpointsResolver interface
+// that uses EDS to resolve the given name to endpoints.
+func newEDSResolver(nameToWatch string, producer xdsresource.Producer, topLevelResolver topLevelResolver) *edsDiscoveryMechanism {
+	ret := &edsDiscoveryMechanism{
+		nameToWatch:      nameToWatch,
+		topLevelResolver: topLevelResolver,
+		stopped:          grpcsync.NewEvent(),
 	}
+	ret.cancelWatch = xdsresource.WatchEndpoints(producer, nameToWatch, ret)
+	return ret
+}
 
-	if err != nil {
-		er.topLevelResolver.onError(err)
+// OnUpdate is invoked to report an update for the resource being watched.
+func (er *edsDiscoveryMechanism) OnUpdate(update *xdsresource.EndpointsResourceData) {
```

Comment:
There is definitely some scope for cleanup here (and your points all related to exiting code). But I don't want to do that in this PR, because there is also a related `dnsDiscoveryMechanism` similar to the `edsDiscoveryMechanism` and they both implement the `endpointsResolver` interface. If I get started on this cleanup, it will make the change much larger and will deviate from what I'm trying to do as part of this PR. If you think this cleanup will help, we can file an issue and track it separately. Thanks.

Reply:
Ah, ok. I guess it makes sense to call it xxxDiscoveryMechanism actually since DiscoveryMechanism from above is what determines what each priority needs to do to get cluster configuration.